File: ocaml/GCBench.ml

type tree = Empty 
          | Node of node
and node = { left: tree ref; right: tree ref; i: int ; j: int }

let kStretchTreeDepth = 18;;
let kLongLivedTreeDepth = 16;;
let kArraySize          = 500000;;
let kMinTreeDepth       = 4;;
let kMaxTreeDepth       = 16;;


let rec treeHelper i = if i = 0 then 1 else 2 * treeHelper(i-1);;

let treeSize i = treeHelper(i+1) - 1;;


let numIters i = (2 * (treeSize kStretchTreeDepth)) / (treeSize i);;

let make_empty_node() = Node { left= ref Empty; right= ref Empty; i= 0; j= 0 }
let make_node (l, r) = Node { left= ref l; right= ref r; i= 0; j= 0 }

(*   
(* needs revision on this populate function *)
let rec populate iDepth node {left=lr; right=rr; i;j}  = 
   if iDepth <= 0 then (make_empty_node())
   else let iDepth = iDepth - 1 
		 in (
             lr := make_empty_node();
             rr := make_empty_node();
			 populate iDepth lr;
			 populate iDepth rr;
		  );;
*)

(* maybe also needs revision on this makeTree function *)
let rec makeTree iDepth =  if iDepth <= 0 then make_empty_node()
					   else make_node ( makeTree (iDepth - 1) , makeTree (iDepth - 1) )



let printDiagnostics () = ( 
	let lTotalMemory, 
		lFreeMemory = 8. *. ( (Gc.stat()).Gc.minor_words +. (Gc.stat()).Gc.major_words ), 
					  8. *. ( (Gc.stat()).Gc.minor_words +. (Gc.stat()).Gc.major_words -. (Gc.stat()).Gc.promoted_words )
	in
	(print_string(" Total memory available="^string_of_float(lTotalMemory)^" bytes");
	 print_string("  Free memory="^string_of_float(lFreeMemory)^" bytes\n") )

(*The total amount of memory allocated by the program since it was started is (in words) minor_words + major_words - promoted_words. Multiply by the word size (4 on a 32-bit machine, 8 on a 64-bit machine) to get the number of bytes.*)
(* Source:
http://caml.inria.fr/pub/docs/manual-ocaml/libref/Gc.html
*)

)


let timeConstruction depth = (
  let iNumIters = numIters depth
  in (
	  print_string("Creating " ^ string_of_int(iNumIters) ^ 
					" trees of depth " ^string_of_int(depth) ^ "\n");
	  let tStart = Sys.time()
(* needs revision on populate....
		   let rec loop i =
				if i < iNumIters then (  populate (depth, make_empty_node); loop (i+1) )
				else ()
			in (loop 0) *)
	  in ( print_string(  "\tTop down construction took "
				     	   ^ string_of_float( (Sys.time() -. tStart) *. 1000. )
					       ^ "msecs\n" )
	     );

      let tStart = Sys.time()

	  in (
			let rec loop i =
				if i < iNumIters then (makeTree depth; loop (i+1) )
				else ()
			in (
			  loop 0;
			 print_string(  "\tBottom up construction took "
			     		  ^ string_of_float( (Sys.time() -. tStart) *. 1000. )
				          ^ "msecs\n" )
			
			   )
		 )
     )
)


let main() = (
  print_string("Garbage Collector Test\n");

  print_string(" Stretching memory with a binary tree of depth " ^
			   string_of_int(kStretchTreeDepth) ^"\n" );
  printDiagnostics();

  let tStart = Sys.time()
  in ( 
		makeTree kStretchTreeDepth;
		print_string(" Creating a long-lived binary tree of depth " ^
		string_of_int(kLongLivedTreeDepth) ^"\n");
		let longLivedTree = make_empty_node()
		in (
      	   (* populate kLongLivedTreeDepth, longLivedTree;  *)

		   );


		print_string(" Creating a long-lived array of " ^
				  	   string_of_int(kArraySize) ^
					   " floats\n");

		let arr = Array.make 500000 0.0
		in ( 
		     print_float (Array.get arr 0); print_string("\n"); 
			 for i=0 to 500000/2 - 1 do ( Array.set arr i (1.0 /. float_of_int(i)) ) done 
		   );

		printDiagnostics();

		for i= kMinTreeDepth/2 to kMaxTreeDepth/2 do timeConstruction (i*2) done;

		if longLivedTree = Empty or (Array.get arr 1000) <> 1.0 /. 1000. then
			print_string("Failed\n")
		else ();

		let tElapsed = Sys.time() -. tStart
		in (
			 printDiagnostics();
			 print_string( "Completed in " ^ 
						   string_of_float( tElapsed *. 1000.) ^ 
						   "ms.\n")
				
		   )
		 
      )

)


let () = main()

File: Erlang/gcBench.erl

-module(gcBench).

-export([nonEmptyNode/2, populate/2, makeTree/1, emptyNode/0,numIters/1,treeSize/1,timeConstruction/1]).
-export([printDiagnostics/0,fillArray/2]).
-export([makeArr/0,kArraySize/0]).
-export([fib/1,simpleLoop/1]).
-export([start_gc_thread/4, gc_func/5]).
-export([start_comp_thread/5, comp_func/6]).


-export([main/1]).



-define(kMinTreeDepth,4).
-define(kMaxTreeDepth,16).
-define(kStretchTreeDepth,18).

-define(EMPTY_NODE, {node, 'empty'}).
%Tree setup

emptyNode () -> ?EMPTY_NODE.

nonEmptyNode(Left,Right) -> {node,{Left,Right}}.


% Build tree bottom-up
makeTree(IDepth) when IDepth > 0 ->
    nonEmptyNode(makeTree(IDepth - 1), makeTree(IDepth - 1));
makeTree(0) ->
	emptyNode ().


%kStretchTreeDepth() -> 18 .
%kLongLivedTreeDepth() -> 16 .
%kMinTreeDepth() -> 4 .
%kMaxTreeDepth() -> 16 .

kArraySize () -> 500000.




% Build tree top down, assigning to older objects.
%populate(IDepth, {node,#{left:=Left , right:=Right} = N}) when IDepth > 0 ->
%        N = N#{left:= emptyNode(), right := emptyNode()},

%	populate(IDepth - 1, maps:get(left,N)),
%	populate(IDepth - 1, maps:get(right,N));
%

populate(IDepth,ThisNode) when IDepth >0 ->

    case ThisNode of
        {node,'empty'} -> {node,{populate(IDepth-1,emptyNode()),populate(IDepth-1,emptyNode())}};
        Otherwise -> io:format("No match:~p~n",[Otherwise])
    end;
populate(0, _) ->
	ok. 

%populate(IDepth,N) when IDepth > 0 ->
%       N = N#{left:= emptyNode(), right:= emptyNode()},

%       populate(IDepth-1, maps:get(left,N)),
%       populate(IDepth-1,maps:get(right,N)).

treeSize(I) -> (1 bsl (I+1) -1).

numIters(I) ->
    (2 * treeSize(?kStretchTreeDepth)) div treeSize(I).

timeConstruction(Depth) -> 
    INumIters = numIters (Depth),
    %io:format("~s~w~s~w~n",["Creating ",INumIters," trees of depth ", Depth]),
    TopDown = fun Loop(CNT)->
                  case CNT < INumIters of
                    true ->  populate(Depth,emptyNode()), Loop(CNT+1);
                    false -> ok
                  end
              end,
    TopDown(0),

    BottomUp = fun Loop(CNT) ->
                  case CNT < INumIters of
                    true -> makeTree(Depth), Loop(CNT+1);
                    false -> ok
                  end
                end,
    BottomUp(0).
    




printDiagnostics() ->

	application:start(sasl),  % have to run with these two lines (just once)
	application:start(os_mon), % to execute OS_MON
	GetData = memsup:get_system_memory_data(),
	{ _, LtotalMemory} = lists:nth(1, GetData),
	{ _, LfreeMemory} = lists:nth(2, GetData),

	io:fwrite(" Total memory available=" ++ integer_to_list(LtotalMemory) ++ " bytes"),
	io:fwrite("  Free memory=" ++ integer_to_list(LfreeMemory) ++ " bytes\n").


%making an empty arrayerlang global variable
makeArr() -> array:new(kArraySize(), {default,0.0}).

fillArray (NewArr,C) when C > 0 ->
    NewVal = 1.0 / (C * 1.0),
    TempArr = array:set(C,NewVal,NewArr),
    fillArray(TempArr,C-1);
fillArray(NewArr,0) ->
    array:set(0, 1.0,NewArr).




fib(0) -> 0 ; 
fib(1) -> 1 ; 
fib(N) when N > 0 -> fib(N-1) + fib(N-2) .

simpleLoop(I) when I > 0->
   simpleLoop(I-1);
simpleLoop(0)-> ok.



start_gc_thread(Num_threads, Tree_depth, Iterations, ID) when Num_threads > 0 ->

	io:format("~s~w~n", ["starting GC thread #", ID] ),

	spawn_opt(gcBench, gc_func, [self(), Tree_depth, ID, Iterations , 0],[{max_heap_size, #{size => 7048608, kill => true, error_logger => true}}]),
	
        start_gc_thread(Num_threads - 1, Tree_depth, Iterations, ID + 1);

start_gc_thread(0, _, _, _) ->
	ok.

gc_func(PID, Tree_depth, ID, Iterations, I) when Iterations > 0 ->

        
        %io:format("~w~n",[process_info(self(),total_heap_size)]),
        Start_time = erlang:timestamp(),
    	
        %stretch the memory space quicklyy
        makeTree(Tree_depth),

        %create a long lived object

        LongLivedTree = emptyNode (),
        populate(Tree_depth,LongLivedTree),
        
        %TempVal = kArraySize()/2,

        LongLivedArray = fillArray(makeArr(),250000),
        
        D = ?kMinTreeDepth,

        Constr = fun While(A) ->
                    case A =< Tree_depth of
                        true -> timeConstruction(A),While(A+2);
                        false -> ok
                    end
                 end,

        Constr(D),
    
        ArrVal =  array:get(1000,LongLivedArray),

        if
            ArrVal /= (1.0/1000) -> io:format("Failed~n");
            true -> ok
        end,
    
	Stop_time = erlang:timestamp(),
        Total_time = timer:now_diff (Stop_time,Start_time),
        io:format("~s~w~s~w~n",["gc:",ID,": Time taken = ",Total_time]),
        
        timer:sleep(3000),

	gc_func(PID, Tree_depth, ID, Iterations - 1 , I + 1 );

gc_func(PID,_,_,0,_) ->
	PID ! done.

				%    1			37       10        1      1
start_comp_thread(Num_threads, Depth, Iterations, ID, Comp_sleep) when Num_threads > 0 ->
	io:format("~s~w~n", ["starting computing thread #", ID] ),

	spawn(gcBench, comp_func, [self(), Depth, ID, Iterations , 0, Comp_sleep] ),

        start_comp_thread(Num_threads - 1, Depth, Iterations, ID + 1, Comp_sleep);

start_comp_thread(0, _, _, _, _) ->
	ok.


comp_func(PID, Depth, ID, Iterations, I, Comp_sleep) when Iterations > 0 ->
	Start_time = erlang:timestamp(),
	%io:format("~s~w~s~w~s~s~n", ["comp:start:",ID,":",I,":",Start_time]),
	fib(Depth),

	Stop_time = erlang:timestamp(),
	%io:format("~s~w~s~w~s~s~n", ["comp:stop:",ID,":",I,":",Stop_time]),
        
       % io:format("~w~n",[process_info(self(),total_heap_size)]),
       % io:format("~w~n",[process_info(self(),memory)]),
        Total_time = timer:now_diff(Stop_time,Start_time),
        io:format("~s~w~s~w~n",["comp:",ID,": Time taken = ",Total_time]),

	timer:sleep(Comp_sleep * 1000),
	comp_func(PID, Depth, ID, Iterations - 1 , I + 1, Comp_sleep);
comp_func(PID,_,_,0,_,_)->
	PID ! done.


main(Args) ->
	_t = lists:nth(1, Args ),
	_d = lists:nth(2, Args ),
	_i = lists:nth(3, Args ),
	_s = lists:nth(4, Args ),
	_g = lists:nth(5, Args ),
	_e = lists:nth(6, Args ),

	{T, _} = string:to_integer( _t ),
	{D, _} = string:to_integer( _d ),
	{I, _} = string:to_integer( _i ),
	{S, _} = string:to_integer( _s ),
	%{G, _} = string:to_integer( _g ),
	%{E, _} = string:to_integer( _e ),
        
        if
		(T > 0 ) ->
			start_comp_thread(T,D,I,1,S);
		true -> ok
	end,

       
	%if
	%	(G > 0 ) ->
	%		start_gc_thread(G,E,I,1);
	%	true -> ok
	%end,

         
	timer:sleep(30000),
	init:stop().




	%% compile
	%erlc gcBench.erl

	%% run
	%erl -noshell -run gcBench main 1 37 10 1 1 10 -s


%
